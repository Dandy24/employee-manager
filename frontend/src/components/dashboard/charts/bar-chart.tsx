import { observer } from 'mobx-react-lite';
import Title from 'antd/lib/typography/Title';
import { Bar, BarChart, CartesianGrid, Legend, ResponsiveContainer, Tooltip, XAxis, YAxis } from 'recharts';
import { Col } from 'antd';
import React from 'react';
import { ExtendedHoursTypeGraphDataInterface } from '../../../models/interfaces/graph-data-interface';
import { useRootStore } from '../../../stores/root-store-provider';

export interface BarChartProps {
    title: string;
    data: ExtendedHoursTypeGraphDataInterface[];
    dataKey1: string;
    dataName1: string;
    dataKey2: string;
    dataName2: string;
    dataKey3?: string;
    dataName3?: string;
    dataKey4?: string;
    dataName4?: string;
    xAxisKey: string;
    clickable?: boolean;
}

export const MyBarChart: React.FC<BarChartProps> = observer((props: BarChartProps) => {
    const {
        title,
        data,
        dataKey1,
        dataName1,
        dataKey2,
        dataName2,
        dataKey3,
        dataName3,
        dataKey4,
        dataName4,
        xAxisKey,
        clickable,
    } = props;

    const rootStore = useRootStore();

    const clickAction = async (chart) => {
        await rootStore.employeeStore.fetchAllEmployees(undefined, undefined, chart.payload.id);
        rootStore.dashboardStore.switchMode();
    };

    return (
        <Col style={{ width: '50%', height: '108%' }}>
            <Title level={2} style={{ textAlign: 'center' }}>
                {title}
            </Title>
            <ResponsiveContainer width="100%" height="85%" id={'bar-chart'}>
                <BarChart
                    width={500}
                    height={300}
                    data={data}
                    margin={{
                        top: 5,
                        right: 30,
                        left: 20,
                        bottom: 5,
                    }}
                    id={'bar-chart'}
                >
                    <CartesianGrid strokeDasharray="4 4" />
                    <XAxis dataKey={xAxisKey} tickFormatter={(name, place) => `${place + 1}. ${name}`} />
                    <YAxis />
                    <Tooltip filterNull={true} formatter={(amount) => `${amount} hod`} />
                    <Legend />
                    <Bar
                        dataKey={dataKey1}
                        stackId="a"
                        name={dataName1}
                        fill="#8884d8"
                        onClick={clickable ? clickAction : null}
                        role={`bar-chart-${dataKey1}-bar`}
                    />
                    <Bar
                        dataKey={dataKey4}
                        stackId="a"
                        name={dataName4}
                        fill="#ff8346"
                        onClick={clickable ? clickAction : null}
                        role={`bar-chart-${dataKey4}-bar`}
                    />
                    {dataKey2 && (
                        <Bar
                            dataKey={dataKey2}
                            name={dataName2}
                            fill="#82ca9d"
                            onClick={clickable ? clickAction : null}
                            role={`bar-chart-${dataKey2}-bar`}
                        />
                    )}
                    {dataKey3 && (
                        <Bar
                            dataKey={dataKey3}
                            name={dataName3}
                            fill="#ffbd37"
                            onClick={clickable ? clickAction : null}
                            role={`bar-chart-${dataKey3}-bar`}
                        />
                    )}
                </BarChart>
            </ResponsiveContainer>
        </Col>
    );
});
