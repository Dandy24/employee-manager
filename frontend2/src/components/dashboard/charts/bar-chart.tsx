import { observer } from 'mobx-react-lite';
import Title from 'antd/lib/typography/Title';
import { Bar, BarChart, CartesianGrid, Legend, ResponsiveContainer, Tooltip, XAxis, YAxis } from 'recharts';
import { Col } from 'antd';
import React from 'react';
import { HoursTypeGraphDataInterface } from '../../../models/interfaces/graph-data-interface';

export interface BarChartProps {
    title: string;
    data: HoursTypeGraphDataInterface[];
    dataKey1: string;
    dataName1: string;
    dataKey2: string;
    dataName2: string;
    dataKey3?: string;
    dataName3?: string;
    dataKey4?: string;
    dataName4?: string;
    xAxisKey: string;
}

export const MyBarChart: React.FC<BarChartProps> = observer((props: BarChartProps) => {
    const {
        title,
        data,
        dataKey1,
        dataName1,
        dataKey2,
        dataName2,
        dataKey3,
        dataName3,
        dataKey4,
        dataName4,
        xAxisKey,
    } = props;

    return (
        <Col style={{ width: '50%', height: '108%' }}>
            <Title level={2} style={{ textAlign: 'center' }}>
                {title}
            </Title>
            <ResponsiveContainer width="100%" height="85%">
                <BarChart
                    width={500}
                    height={300}
                    data={data}
                    margin={{
                        top: 5,
                        right: 30,
                        left: 20,
                        bottom: 5,
                    }}
                >
                    <CartesianGrid strokeDasharray="3 3" />
                    <XAxis dataKey={xAxisKey} />
                    <YAxis />
                    <Tooltip />
                    <Legend />
                    <Bar dataKey={dataKey1} stackId="a" name={dataName1} fill="#8884d8" />
                    <Bar dataKey={dataKey4} stackId="a" name={dataName4} fill="#ff8346" />
                    <Bar dataKey={dataKey2} name={dataName2} fill="#82ca9d" />
                    {dataKey3 && <Bar dataKey={dataKey3} name={dataName3} fill="#ffbd37" />}
                </BarChart>
            </ResponsiveContainer>
        </Col>
    );
});
