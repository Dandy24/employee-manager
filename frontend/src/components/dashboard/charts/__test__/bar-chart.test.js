import { render } from '@testing-library/react';
import React from 'react';
import { MyBarChart } from '../bar-chart';

jest.mock('recharts', () => {
    const OriginalModule = jest.requireActual('recharts');

    return {
        ...OriginalModule,
        ResponsiveContainer: ({ height, children }) => (
            <OriginalModule.ResponsiveContainer width={800} height={500}>
                {children}
            </OriginalModule.ResponsiveContainer>
        ),
    };
});

test('Bar chart matches snapshot', async () => {
    Object.defineProperty(window, 'matchMedia', {
        writable: true,
        value: jest.fn().mockImplementation((query) => ({
            matches: false,
            media: query,
            onchange: null,
            addListener: jest.fn(), // deprecated
            removeListener: jest.fn(), // deprecated
            addEventListener: jest.fn(),
            removeEventListener: jest.fn(),
            dispatchEvent: jest.fn(),
        })),
    });

    const testData = [
        { id: 1, name: 'Test', sick: 2, work: 160, vac: 6 },
        { id: 2, name: 'Test2', sick: 10, work: 150, vac: 16 },
    ];

    const view = render(
        <MyBarChart
            title="Nejlepší měsíce podle výkonu"
            data={testData}
            dataKey1="work"
            dataName1="V práci"
            dataKey2="vac"
            dataName2="Volno"
            xAxisKey="name"
        />,
    );

    // eslint-disable-next-line testing-library/no-node-access,testing-library/no-container
    const barChart = view.container.querySelector('#bar-chart');

    expect(barChart).toMatchSnapshot();
});
