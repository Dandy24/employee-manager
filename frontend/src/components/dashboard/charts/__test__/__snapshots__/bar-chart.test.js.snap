// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Bar chart matches snapshot 1`] = `
<svg
  class="recharts-surface"
  height="500"
  id="bar-chart"
  version="1.1"
  viewBox="0 0 800 500"
  width="800"
>
  <defs>
    <clippath
      id="bar-chart-clip"
    >
      <rect
        height="460"
        width="690"
        x="80"
        y="5"
      />
    </clippath>
  </defs>
  <g
    class="recharts-cartesian-grid"
  >
    <g
      class="recharts-cartesian-grid-horizontal"
    >
      <line
        fill="none"
        height="460"
        offset="[object Object]"
        stroke="#ccc"
        stroke-dasharray="4 4"
        width="690"
        x="80"
        x1="80"
        x2="770"
        y="5"
        y1="465"
        y2="465"
      />
      <line
        fill="none"
        height="460"
        offset="[object Object]"
        stroke="#ccc"
        stroke-dasharray="4 4"
        width="690"
        x="80"
        x1="80"
        x2="770"
        y="5"
        y1="350"
        y2="350"
      />
      <line
        fill="none"
        height="460"
        offset="[object Object]"
        stroke="#ccc"
        stroke-dasharray="4 4"
        width="690"
        x="80"
        x1="80"
        x2="770"
        y="5"
        y1="235"
        y2="235"
      />
      <line
        fill="none"
        height="460"
        offset="[object Object]"
        stroke="#ccc"
        stroke-dasharray="4 4"
        width="690"
        x="80"
        x1="80"
        x2="770"
        y="5"
        y1="120"
        y2="120"
      />
      <line
        fill="none"
        height="460"
        offset="[object Object]"
        stroke="#ccc"
        stroke-dasharray="4 4"
        width="690"
        x="80"
        x1="80"
        x2="770"
        y="5"
        y1="5"
        y2="5"
      />
    </g>
    <g
      class="recharts-cartesian-grid-vertical"
    >
      <line
        fill="none"
        height="460"
        offset="[object Object]"
        stroke="#ccc"
        stroke-dasharray="4 4"
        width="690"
        x="80"
        x1="252.5"
        x2="252.5"
        y="5"
        y1="5"
        y2="465"
      />
      <line
        fill="none"
        height="460"
        offset="[object Object]"
        stroke="#ccc"
        stroke-dasharray="4 4"
        width="690"
        x="80"
        x1="597.5"
        x2="597.5"
        y="5"
        y1="5"
        y2="465"
      />
      <line
        fill="none"
        height="460"
        offset="[object Object]"
        stroke="#ccc"
        stroke-dasharray="4 4"
        width="690"
        x="80"
        x1="80"
        x2="80"
        y="5"
        y1="5"
        y2="465"
      />
      <line
        fill="none"
        height="460"
        offset="[object Object]"
        stroke="#ccc"
        stroke-dasharray="4 4"
        width="690"
        x="80"
        x1="770"
        x2="770"
        y="5"
        y1="5"
        y2="465"
      />
    </g>
  </g>
  <g
    class="recharts-layer recharts-cartesian-axis recharts-xAxis xAxis"
  >
    <line
      class="recharts-cartesian-axis-line"
      fill="none"
      height="30"
      orientation="bottom"
      stroke="#666"
      type="category"
      width="690"
      x="80"
      x1="80"
      x2="770"
      y="465"
      y1="465"
      y2="465"
    />
    <g
      class="recharts-cartesian-axis-ticks"
    >
      <g
        class="recharts-layer recharts-cartesian-axis-tick"
      >
        <line
          class="recharts-cartesian-axis-tick-line"
          fill="none"
          height="30"
          orientation="bottom"
          stroke="#666"
          type="category"
          width="690"
          x="80"
          x1="252.5"
          x2="252.5"
          y="465"
          y1="471"
          y2="465"
        />
        <text
          class="recharts-text recharts-cartesian-axis-tick-value"
          fill="#666"
          height="30"
          orientation="bottom"
          stroke="none"
          text-anchor="middle"
          type="category"
          width="690"
          x="252.5"
          y="473"
        >
          <tspan
            dy="0.71em"
            x="252.5"
          >
            1. Test
          </tspan>
        </text>
      </g>
      <g
        class="recharts-layer recharts-cartesian-axis-tick"
      >
        <line
          class="recharts-cartesian-axis-tick-line"
          fill="none"
          height="30"
          orientation="bottom"
          stroke="#666"
          type="category"
          width="690"
          x="80"
          x1="597.5"
          x2="597.5"
          y="465"
          y1="471"
          y2="465"
        />
        <text
          class="recharts-text recharts-cartesian-axis-tick-value"
          fill="#666"
          height="30"
          orientation="bottom"
          stroke="none"
          text-anchor="middle"
          type="category"
          width="690"
          x="597.5"
          y="473"
        >
          <tspan
            dy="0.71em"
            x="597.5"
          >
            2. Test2
          </tspan>
        </text>
      </g>
    </g>
  </g>
  <g
    class="recharts-layer recharts-cartesian-axis recharts-yAxis yAxis"
  >
    <line
      class="recharts-cartesian-axis-line"
      fill="none"
      height="460"
      orientation="left"
      stroke="#666"
      type="number"
      width="60"
      x="20"
      x1="80"
      x2="80"
      y="5"
      y1="5"
      y2="465"
    />
    <g
      class="recharts-cartesian-axis-ticks"
    >
      <g
        class="recharts-layer recharts-cartesian-axis-tick"
      >
        <line
          class="recharts-cartesian-axis-tick-line"
          fill="none"
          height="460"
          orientation="left"
          stroke="#666"
          type="number"
          width="60"
          x="20"
          x1="74"
          x2="80"
          y="5"
          y1="465"
          y2="465"
        />
        <text
          class="recharts-text recharts-cartesian-axis-tick-value"
          fill="#666"
          height="460"
          orientation="left"
          stroke="none"
          text-anchor="end"
          type="number"
          width="60"
          x="72"
          y="465"
        >
          <tspan
            dy="0.355em"
            x="72"
          >
            0
          </tspan>
        </text>
      </g>
      <g
        class="recharts-layer recharts-cartesian-axis-tick"
      >
        <line
          class="recharts-cartesian-axis-tick-line"
          fill="none"
          height="460"
          orientation="left"
          stroke="#666"
          type="number"
          width="60"
          x="20"
          x1="74"
          x2="80"
          y="5"
          y1="350"
          y2="350"
        />
        <text
          class="recharts-text recharts-cartesian-axis-tick-value"
          fill="#666"
          height="460"
          orientation="left"
          stroke="none"
          text-anchor="end"
          type="number"
          width="60"
          x="72"
          y="350"
        >
          <tspan
            dy="0.355em"
            x="72"
          >
            40
          </tspan>
        </text>
      </g>
      <g
        class="recharts-layer recharts-cartesian-axis-tick"
      >
        <line
          class="recharts-cartesian-axis-tick-line"
          fill="none"
          height="460"
          orientation="left"
          stroke="#666"
          type="number"
          width="60"
          x="20"
          x1="74"
          x2="80"
          y="5"
          y1="235"
          y2="235"
        />
        <text
          class="recharts-text recharts-cartesian-axis-tick-value"
          fill="#666"
          height="460"
          orientation="left"
          stroke="none"
          text-anchor="end"
          type="number"
          width="60"
          x="72"
          y="235"
        >
          <tspan
            dy="0.355em"
            x="72"
          >
            80
          </tspan>
        </text>
      </g>
      <g
        class="recharts-layer recharts-cartesian-axis-tick"
      >
        <line
          class="recharts-cartesian-axis-tick-line"
          fill="none"
          height="460"
          orientation="left"
          stroke="#666"
          type="number"
          width="60"
          x="20"
          x1="74"
          x2="80"
          y="5"
          y1="120"
          y2="120"
        />
        <text
          class="recharts-text recharts-cartesian-axis-tick-value"
          fill="#666"
          height="460"
          orientation="left"
          stroke="none"
          text-anchor="end"
          type="number"
          width="60"
          x="72"
          y="120"
        >
          <tspan
            dy="0.355em"
            x="72"
          >
            120
          </tspan>
        </text>
      </g>
      <g
        class="recharts-layer recharts-cartesian-axis-tick"
      >
        <line
          class="recharts-cartesian-axis-tick-line"
          fill="none"
          height="460"
          orientation="left"
          stroke="#666"
          type="number"
          width="60"
          x="20"
          x1="74"
          x2="80"
          y="5"
          y1="5"
          y2="5"
        />
        <text
          class="recharts-text recharts-cartesian-axis-tick-value"
          fill="#666"
          height="460"
          orientation="left"
          stroke="none"
          text-anchor="end"
          type="number"
          width="60"
          x="72"
          y="5"
        >
          <tspan
            dy="0.355em"
            x="72"
          >
            160
          </tspan>
        </text>
      </g>
    </g>
  </g>
  <g
    class="recharts-layer recharts-bar"
  >
    <g
      class="recharts-layer recharts-bar-rectangles"
    >
      <g
        class="recharts-layer"
      >
        <g
          class="recharts-layer recharts-bar-rectangle"
        />
        <g
          class="recharts-layer recharts-bar-rectangle"
        />
      </g>
    </g>
  </g>
  <g
    class="recharts-layer recharts-bar"
  >
    <g
      class="recharts-layer recharts-bar-rectangles"
    >
      <g
        class="recharts-layer"
      >
        <g
          class="recharts-layer recharts-bar-rectangle"
        />
        <g
          class="recharts-layer recharts-bar-rectangle"
        />
      </g>
    </g>
  </g>
  <g
    class="recharts-layer recharts-bar"
  >
    <g
      class="recharts-layer recharts-bar-rectangles"
    >
      <g
        class="recharts-layer"
      >
        <g
          class="recharts-layer recharts-bar-rectangle"
        />
        <g
          class="recharts-layer recharts-bar-rectangle"
        />
      </g>
    </g>
  </g>
</svg>
`;
